benches/blob_store.rs

```diff
@@ -0,0 +1,171 @@
+    for (index, b) in blobs.iter_mut().enumerate() {
+        b.set_index(index as u64).unwrap();
+        b.set_slot(slot).unwrap();
+    }
+
+    store
+        .put_blobs(&blobs)
+        .expect("Expectd successful insertion of blobs into ledger");
+}
+
+// Write small blobs to the ledger
+#[bench]
+#[ignore]
+fn bench_write_small(bench: &mut Bencher) {
+    let ledger_path = get_tmp_store_path("bench_write_small").unwrap();
+    let num_entries = 32 * 1024;
+    let entries = make_tiny_test_entries(num_entries);
+    let mut blobs = entries.to_blobs();
+    for (index, b) in blobs.iter_mut().enumerate() {
+        b.set_index(index as u64).unwrap();
+    }
+    bench_write_blobs(bench, &mut blobs, &ledger_path.to_string_lossy());
+}
+
+// Write big blobs to the ledger
+#[bench]
+#[ignore]
+fn bench_write_big(bench: &mut Bencher) {
+    let ledger_path = get_tmp_store_path("bench_write_big").unwrap();
+    let num_entries = 1 * 1024;
+    let entries = make_large_test_entries(num_entries);
+    let mut blobs = entries.to_blobs();
+    for (index, b) in blobs.iter_mut().enumerate() {
+        b.set_index(index as u64).unwrap();
+    }
+
+    bench_write_blobs(bench, &mut blobs, &ledger_path.to_string_lossy());
+}
+
+#[bench]
+#[ignore]
+fn bench_read_sequential(bench: &mut Bencher) {
+    let ledger_path = get_tmp_store_path("bench_read_sequential").unwrap();
+    let mut store = Store::open(&ledger_path);
+
+    // Insert some big and small blobs into the ledger
+    let num_small_blobs = 32 * 1024;
+    let num_large_blobs = 32 * 1024;
+    let total_blobs = num_small_blobs + num_large_blobs;
+    let slot = 0;
+    setup_read_bench(&mut store, num_small_blobs, num_large_blobs, slot);
+
+    let num_reads = total_blobs / 15;
+    let mut rng = rand::thread_rng();
+    bench.iter(move || {
+        // Generate random starting point in the range [0, total_blobs - 1], read num_reads blobs sequentially
+        let start_index = rng.gen_range(0, num_small_blobs + num_large_blobs);
+        for i in start_index..start_index + num_reads {
+            let _ = store.get_blob_data(slot, i as u64 % total_blobs);
+        }
+    });
+
+    Store::destroy(&ledger_path).expect("Expected successful database destruction");
+}
+
+#[bench]
+#[ignore]
+fn bench_read_random(bench: &mut Bencher) {
+    let ledger_path = get_tmp_store_path("bench_read_random").unwrap();
+    let mut store = Store::open(&ledger_path);
+
+    // Insert some big and small blobs into the ledger
+    let num_small_blobs = 32 * 1024;
+    let num_large_blobs = 32 * 1024;
+    let total_blobs = num_small_blobs + num_large_blobs;
+    let slot = 0;
+    setup_read_bench(&mut store, num_small_blobs, num_large_blobs, slot);
+
+    let num_reads = total_blobs / 15;
+
+    // Generate a num_reads sized random sample of indexes in range [0, total_blobs - 1],
+    // simulating random reads
+    let mut rng = rand::thread_rng();
+    let indexes: Vec<usize> = (0..num_reads)
```

Comment:
are you intending to verify that caching helps, once implemented?

Reply:
Yes. I've also experimented with parallelizing writing (partitioned based on slots) but I couldn't get it to work safely without needing to copy so much memory that it was much slower overall.

The little caching I've done so far made benchmark ns/iter ~70-80% of  `db_ledger` benchmarks when before it was about ~110-120% . 

I tried several different ways of exploiting concurrency including using `tokio` and futures, creating a separate writer thread that communicated over `std::sync::mpsc` and `crossbeam-channel::` channels, etc. 

But I could not do it both 1) safely 2) without so much copying that it made everything much slower.